### ISSUE TITLE
Multi-Version CRDs Can Break When Admission Webhooks Are Registered

### ISSUE THREAD

**Issue by maxsmythe** [kubernetes/kubernetes#80494]
**What happened**:

Multi-version CRDs can break when the stored version changes and a user attempts to re-write a previously saved CR represented using the (newer) stored version.

This only happens when an admission webhook is registered (and, I'm guessing, the webhook matches the request), regardless of whether the webhook's server is running or not.

Impact seen here: https://github.com/open-policy-agent/gatekeeper/pull/190#discussion_r306511055

**What you expected to happen**:

I would expect that multi-version CRDs with a version-conversion strategy would be able to accept a write using any version.

**How to reproduce it (as minimally and precisely as possible)**:

1) Install any validating webhook (e.g. Gatekeeper) that listens to the created CRD

2) Create a CRD:

```yaml
apiVersion: apiextensions.k8s.io/v1beta1
kind: CustomResourceDefinition
metadata:
  name: tests.testgroup.com
spec:
  group: testgroup.com
  names:
    kind: Test
    listKind: TestList
    plural: tests
    singular: test
  scope: Cluster
  version: v1alpha1
  versions:
  - name: v1alpha1
    served: true
    storage: true
```

3) Create a Resource:

```yaml
apiVersion: testgroup.com/v1alpha1
kind: Test
metadata:
  name: test
spec:
  some: thing2
```

4) Update the CRD:

```yaml
apiVersion: apiextensions.k8s.io/v1beta1
kind: CustomResourceDefinition
metadata:
  name: tests.testgroup.com
spec:
  group: testgroup.com
  names:
    kind: Test
    listKind: TestList
    plural: tests
    singular: test
  scope: Cluster
  version: v1beta1
  versions:
  - name: v1beta1
    served: true
    storage: true
  - name: v1alpha1
    served: true
    storage: false
```

5) Try to apply a resource:

```yaml
apiVersion: testgroup.com/v1beta1
kind: Test
metadata:
  name: test
```

See error:

```
Error from server (InternalError): error when applying patch:
{"metadata":{"annotations":{"kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"testgroup.com/v1beta1\",\"kind\":\"Test\",\"metadata\":{\"annotations\":{},\"name\":\"test\"}}\n"}},"spec":null}
to:
Resource: "testgroup.com/v1beta1, Resource=tests", GroupVersionKind: "testgroup.com/v1beta1, Kind=Test"
Name: "test", Namespace: ""
Object: &{map["apiVersion":"testgroup.com/v1beta1" "kind":"Test" "metadata":map["annotations":map["kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"testgroup.com/v1alpha1\",\"kind\":\"Test\",\"metadata\":{\"annotations\":{},\"name\":\"test\"},\"spec\":{\"some\":\"thing2\"}}\n"] "creationTimestamp":"2019-07-23T10:34:00Z" "generation":'\x01' "name":"test" "resourceVersion":"118715" "selfLink":"/apis/testgroup.com/v1beta1/tests/test" "uid":"624f296a-ad35-11e9-9d16-42010a800100"] "spec":map["some":"thing2"]]}
for: "test.yaml": Internal error occurred: no kind "Test" is registered for version "testgroup.com/v1beta1" in scheme "k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go:52"
```

**Anything else we need to know?**:

Here is the log output from my API server:

```
I0723 21:47:11.584460       1 wrap.go:47] PATCH /apis/testgroup.com/v1beta1/tests/test: (63.064558ms) 500
goroutine 10796360 [running]:
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/httplog.(*respLogger).recordStatus(0xc00795b650, 0x1f4)
        /workspace/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/httplog/httplog.go:204 +0xd2
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/httplog.(*respLogger).WriteHeader(0xc00795b650, 0x1f4)
        /workspace/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/httplog/httplog.go:183 +0x35
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.(*baseTimeoutWriter).WriteHeader(0xc000721ae0, 0x1f4)
        /workspace/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/timeout.go:205 +0xaf
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.(*auditResponseWriter).WriteHeader(0xc006a02000, 0x1f4)
        /workspace/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters/audit.go:219 +0x63
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/metrics.(*ResponseWriterDelegator).WriteHeader(0xc00872f320, 0x1f4)
        /workspace/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:319 +0x45
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters.httpResponseWriterWithInit.Write(0x0, 0x3febf6e, 0x10, 0x1f4, 0x5ee7360, 0xc00627eaa0, 0xc013d82000, 0x1d3, 0x21d74, 0x89e0100, ...)
        /workspace/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go:51 +0xfd
encoding/json.(*Encoder).Encode(0xc00d4a3a18, 0x3eac900, 0xc008f8cab0, 0x6, 0x0)
        /usr/local/go/src/encoding/json/stream.go:225 +0x1ca
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime/serializer/json.(*Serializer).Encode(0xc011a60e80, 0x5ecd080, 0xc008f8cab0, 0x5ebe680, 0xc00de1a090, 0x3226496, 0x6)
        /workspace/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go:246 +0x2eb
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime/serializer/versioning.(*codec).Encode(0xc008f8cbd0, 0x5ecd080, 0xc008f8cab0, 0x5ebe680, 0xc00de1a090, 0xc008f8cbd0, 0x5ee5de0)
        /workspace/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime/serializer/versioning/versioning.go:211 +0x274
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters.SerializeObject(0x3febf6e, 0x10, 0x7f97a715a9b0, 0xc008f8cbd0, 0x5ee7360, 0xc00627eaa0, 0xc00ab88d00, 0x1f4, 0x5ecd080, 0xc008f8cab0)
        /workspace/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go:116 +0x11b
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters.WriteObjectNegotiated(0x5ee9ce0, 0xc008d63140, 0xc010d3baa0, 0xd, 0xc010d3bad7, 0x7, 0x5ee7360, 0xc00627eaa0, 0xc00ab88d00, 0x1f4, ...)
        /workspace/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go:142 +0x370
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters.ErrorNegotiated(0x5eba720, 0xc008f8ca20, 0x5ee9ce0, 0xc008d63140, 0xc010d3baa0, 0xd, 0xc010d3bad7, 0x7, 0x5ee7360, 0xc00627eaa0, ...)
        /workspace/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go:161 +0x149
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers.(*RequestScope).err(0xc0088a1080, 0x5eba720, 0xc008f8ca20, 0x5ee7360, 0xc00627eaa0, 0xc00ab88d00)
        /workspace/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/rest.go:80 +0xdd
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers.PatchResource.func1(0x5ee7360, 0xc00627eaa0, 0xc00ab88d00)
        /workspace/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/patch.go:193 +0x13e6
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/metrics.InstrumentHandlerFunc.func1(0x5ee72a0, 0xc00627ea60, 0xc00ab88d00)
        /workspace/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:250 +0x1ee
k8s.io/kubernetes/vendor/k8s.io/apiextensions-apiserver/pkg/apiserver.(*crdHandler).ServeHTTP(0xc000fd4600, 0x5ee72a0, 0xc00627ea60, 0xc00ab88d00)
        /workspace/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:252 +0x4b3
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/mux.(*pathHandler).ServeHTTP(0xc0013912c0, 0x5ee72a0, 0xc00627ea60, 0xc00ab88d00)
        /workspace/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/mux/pathrecorder.go:248 +0x394
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/mux.(*PathRecorderMux).ServeHTTP(0xc00064b490, 0x5ee72a0, 0xc00627ea60, 0xc00ab88d00)
        /workspace/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/mux/pathrecorder.go:234 +0x8a
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server.director.ServeHTTP(0x40073e2, 0x17, 0xc0009501b0, 0xc00064b490, 0x5ee72a0, 0xc00627ea60, 0xc00ab88d00)
        /workspace/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/handler.go:154 +0x661
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/mux.(*pathHandler).ServeHTTP(0xc003ec3e80, 0x5ee72a0, 0xc00627ea60, 0xc00ab88d00)
        /workspace/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/mux/pathrecorder.go:254 +0x22c
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/mux.(*PathRecorderMux).ServeHTTP(0xc000ff3dc0, 0x5ee72a0, 0xc00627ea60, 0xc00ab88d00)
        /workspace/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/mux/pathrecorder.go:234 +0x8a
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server.director.ServeHTTP(0x3fe5f77, 0xe, 0xc000950360, 0xc000ff3dc0, 0x5ee72a0, 0xc00627ea60, 0xc00ab88d00)
        /workspace/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/handler.go:154 +0x661
k8s.io/kubernetes/vendor/k8s.io/kube-aggregator/pkg/apiserver.(*proxyHandler).ServeHTTP(0xc0109cc070, 0x5ee72a0, 0xc00627ea60, 0xc00ab88d00)
        /workspace/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go:90 +0x16a
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/mux.(*pathHandler).ServeHTTP(0xc00348cec0, 0x5ee72a0, 0xc00627ea60, 0xc00ab88d00)
        /workspace/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/mux/pathrecorder.go:248 +0x394
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/mux.(*PathRecorderMux).ServeHTTP(0xc0022af340, 0x5ee72a0, 0xc00627ea60, 0xc00ab88d00)
        /workspace/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/mux/pathrecorder.go:234 +0x8a
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server.director.ServeHTTP(0x3fe908a, 0xf, 0xc003b00000, 0xc0022af340, 0x5ee72a0, 0xc00627ea60, 0xc00ab88d00)
        /workspace/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/handler.go:154 +0x661
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithAuthorization.func1(0x5ee72a0, 0xc00627ea60, 0xc00ab88d00)
        /workspace/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters/authorization.go:64 +0x4c3
net/http.HandlerFunc.ServeHTTP(0xc002a19200, 0x5ee72a0, 0xc00627ea60, 0xc00ab88d00)
        /usr/local/go/src/net/http/server.go:1964 +0x44
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.WithMaxInFlightLimit.func1(0x5ee72a0, 0xc00627ea60, 0xc00ab88d00)
        /workspace/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/maxinflight.go:160 +0x3ff
net/http.HandlerFunc.ServeHTTP(0xc0026dac30, 0x5ee72a0, 0xc00627ea60, 0xc00ab88d00)
        /usr/local/go/src/net/http/server.go:1964 +0x44
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithImpersonation.func1(0x5ee72a0, 0xc00627ea60, 0xc00ab88d00)
        /workspace/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters/impersonation.go:50 +0x1eeb
net/http.HandlerFunc.ServeHTTP(0xc002a19240, 0x5ee72a0, 0xc00627ea60, 0xc00ab88d00)
        /usr/local/go/src/net/http/server.go:1964 +0x44
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithAudit.func1(0x7f979bcd28b0, 0xc00627ea58, 0xc00ab88c00)
        /workspace/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters/audit.go:110 +0x4bc
net/http.HandlerFunc.ServeHTTP(0xc002a19280, 0x7f979bcd28b0, 0xc00627ea58, 0xc00ab88c00)
        /usr/local/go/src/net/http/server.go:1964 +0x44
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithAuthentication.func1(0x7f979bcd28b0, 0xc00627ea58, 0xc00ab88a00)
        /workspace/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters/authentication.go:81 +0x456
net/http.HandlerFunc.ServeHTTP(0xc006a03310, 0x7f979bcd28b0, 0xc00627ea58, 0xc00ab88a00)
        /usr/local/go/src/net/http/server.go:1964 +0x44
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.(*timeoutHandler).ServeHTTP.func1(0xc013eb41e0, 0xc001e9d300, 0x5ef23a0, 0xc00627ea58, 0xc00ab88a00)
        /workspace/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/timeout.go:108 +0xb3
created by k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.(*timeoutHandler).ServeHTTP
        /workspace/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/timeout.go:97 +0x1b0

logging error output: "{\"kind\":\"Status\",\"apiVersion\":\"v1\",\"metadata\":{},\"status\":\"Failure\",\"message\":\"Internal error occurred: no kind \\\"Test\\\" is registered for version \\\"testgroup.com/v1beta1\\\" in scheme \\\"k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go:52\\\"\",\"reason\":\"InternalError\",\"details\":{\"causes\":[{\"message\":\"no kind \\\"Test\\\" is registered for version \\\"testgroup.com/v1beta1\\\" in scheme \\\"k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go:52\\\"\"}]},\"code\":500}\n"
 [kubectl/v1.15.1 (linux/amd64) kubernetes/4485c6f 104.132.11.93:38412]
```

I'm not familiar with debugging the API server, but I took a stab. I believe this is the problem bit of code (taken from master, not the 1.13 branch):

https://github.com/kubernetes/kubernetes/blob/3ea28073a48dc8cb725536b9433b98ea4c55c5f6/staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/generic/conversion.go#L26

In my test version of k8s (1.13.7) the error appears to be getting triggered because the CRD doesn't get upversioned when read from etcd (only version-opinionated CRDs get converted): https://github.com/kubernetes/kubernetes/blob/3ea28073a48dc8cb725536b9433b98ea4c55c5f6/staging/src/k8s.io/apimachinery/pkg/runtime/serializer/versioning/versioning.go#L94)

The workaround in my scenario is to update the CR using the same version as the stored version. The object will get converted in storage and then will work. This appears to be because the default version webhooks try to convert to is provided by the CRD server and is equal to the requested version, which has the effect of not invoking a schema conversion in the webhook plugin.

**Environment**:
- Kubernetes version (use `kubectl version`): v1.13.7-gke.8
- Cloud provider or hardware configuration: Google/GKE
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

**Comment by liggitt** [comment]
This was a known issue fixed in 1.14+ already

See fix in 1.13 in https://github.com/kubernetes/kubernetes/pull/79495

**Comment by maxsmythe** [comment]
ack, thanks for looking, apologies for the duplicate issue.
